### PR TITLE
ARROW-17481: [C++] [Python] Major performance improvements to CSV reading from S3

### PR DESCRIPTION
**Pull request by marsupialtail**
Draft.

### CODE REVIEW

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:549]: I'm not sure this makes sense as a default parameter.  A reader probably wouldn't intuitively guess this.  Let's just force it to be specified.

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:981]: I think this should actually go in `async_generator.h`.  I think I told you otherwise once before but I was thinking it was going to be more tailored to the CSV reader.  This seems pretty generic.  Also, it should ideally be independently unit tested with a few tests.

File: cpp/src/arrow/io/interfaces.h
Comment (westonpace) [inline cpp/src/arrow/io/interfaces.h:348]: We will want to document this to point out that it is async reentrant (e.g. can be read with a readahead generator).  Or, maybe this function could apply the readahead generator also?  So it would be:

```
 Result<AsyncGenerator<std::shared_ptr<Buffer>>> MakeRandomAccessFileGenerator(std::shared_ptr<RandomAccessFile> file, int64_t block_size, int readahead);
```

File: cpp/src/arrow/dataset/file_csv.cc
Comment (westonpace) [inline cpp/src/arrow/dataset/file_csv.cc:211]: Can you create a follow-up JIRA to remove this double-read?  I don't think we need to worry about it now but I'd like to get away from it at some point.

**Review by westonpace — CHANGES_REQUESTED** [review]
Since this is draft these are just high level thoughts.

File: cpp/src/arrow/io/interfaces.h
Comment (marsupialtail) [reply cpp/src/arrow/io/interfaces.h:348]: I think it makes more sense to separate out the readahead generator if you are going to do this. I personally think a lot of interfaces in Arrow already have too many flags and it's not very easy for a new person to figure out what are the good flags to use. I don't think having this generator itself do the readahead will lead to too many performance benefits either?

File: cpp/src/arrow/dataset/file_csv.cc
Comment (marsupialtail) [reply cpp/src/arrow/dataset/file_csv.cc:211]: Yes i plan to do that. This double read is very annoying. The whole logic around what to do with the first block needs to be streamlined and optimized.

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:23]:
```suggestion
```

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:549]: I wonder if this might be more readable as `InputType` instead of `T`.

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:979]: `TonyReader` has a nice ring to it but I think `StreamingFileReader` might be easier to understand for future readers.

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:1011]: Is this the only method that differs from `StreamingReaderImpl`?  Is there any way the two streaming readers could extend some kind of `StreamingReaderBase` to avoid duplication?

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:1316]: I'll defer to the input of others but I think an overload of `MakeStreamingReader` based solely on the type of `input` is a rather subtle API.  Could we maybe just rename this to `MakeStreamingFileReader`?

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:1460]: I suppose the same comment I had for `MakeStreamingReader` applies here too but `MakeFileAsync` doesn't really make sense so maybe we can leave this method name alone and just live with an overload but I'm open to suggestions.

File: cpp/src/arrow/csv/reader.h
Comment (westonpace) [inline cpp/src/arrow/csv/reader.h:105]: Let's add at least a brief comment here explaining the difference between the two reader types.

File: cpp/src/arrow/dataset/file_csv.cc
Comment (westonpace) [inline cpp/src/arrow/dataset/file_csv.cc:21]:
```suggestion
```

File: cpp/src/arrow/io/interfaces.cc
Comment (westonpace) [inline cpp/src/arrow/io/interfaces.cc:22]:
```suggestion
```

File: cpp/src/arrow/io/interfaces.cc
Comment (westonpace) [inline cpp/src/arrow/io/interfaces.cc:146]:
```suggestion
    Status Init() {
```

File: cpp/src/arrow/io/interfaces.cc
Comment (westonpace) [inline cpp/src/arrow/io/interfaces.cc:150]:
```suggestion
      return file->GetSize();
```

You are probably right but I'm not sure avoiding this if check is worth making this method slightly more confusing given it shouldn't really be called all that often.

File: cpp/src/arrow/util/async_generator.h
Comment (westonpace) [inline cpp/src/arrow/util/async_generator.h:151]: Can you expand on this description?  I'm pretty sure the important part is that you are launching a new thread task with each input right?

File: cpp/src/arrow/util/async_generator.h
Comment (westonpace) [inline cpp/src/arrow/util/async_generator.h:188]:
```suggestion
            return state->apply_fun(next);
```
I'm not entirely sure why you are rewriting the failure to "not implemented".

**Review by westonpace — CHANGES_REQUESTED** [review]
A few thoughts but thanks for figuring this out.

File: cpp/src/arrow/csv/reader.cc
Comment (marsupialtail) [reply cpp/src/arrow/csv/reader.cc:979]: Ah thanks for pointing this out. I forgot to change this!! :-P

File: cpp/src/arrow/csv/reader.cc
Comment (marsupialtail) [reply cpp/src/arrow/csv/reader.cc:1011]: The hope is that I will keep improving this TonyReaderImpl (or whatever I change the name to) to include more and more optimizations in subsequent PRs. E.g. parallel decode. So the amount of duplicated code is hopefully a temporary state.

File: cpp/src/arrow/util/async_generator.h
Comment (marsupialtail) [reply cpp/src/arrow/util/async_generator.h:188]: As opposed to just finishing the generator? It needs to return a not ok status here, the error message should be fleshed out more. If it simply finishes instead of returning a not ok, some of the tests will fail. e.g. if you read invalid csv, the parse task would return a not ok status that needs to be reflected up this generator to result in a status not ok for dataset inspect or the like.

File: cpp/src/arrow/util/async_generator.h
Comment (marsupialtail) [reply cpp/src/arrow/util/async_generator.h:151]: That is right.

File: cpp/src/arrow/util/async_generator.h
Comment (westonpace) [reply cpp/src/arrow/util/async_generator.h:188]: If the return value of `state->apply_fun(next)` is a not ok status then `return state->apply_fun(next)` will return a not ok status right?

**Comment by marsupialtail** [comment]
failed CI has nothing to do with this PR. I am going to wait for this one to merge before proceeding with next PR that streamlines logic around first block and parallel decode.

File: cpp/src/arrow/csv/reader.cc
Comment (marsupialtail) [reply cpp/src/arrow/csv/reader.cc:1313]: I think it's ok :-) these are all internal APIs.

File: cpp/src/arrow/io/interfaces.h
Comment (marsupialtail) [reply cpp/src/arrow/io/interfaces.h]: Added a comment

File: cpp/src/arrow/util/async_generator.h
Comment (westonpace) [inline cpp/src/arrow/util/async_generator.h:152]:
```suggestion
/// \brief Similar to MapGenerator but applies the map function in a new thread task
/// This is similar to combining a map generator and transfer generator but the former
/// would not be able to guarantee the map task runs on a new thread.
```

File: cpp/src/arrow/io/interfaces.cc
Comment (westonpace) [inline cpp/src/arrow/io/interfaces.cc:168]:
```suggestion
    // pos is guaranteed to be smaller than total size, but you might not be able to read
    // a full block
```

File: cpp/src/arrow/dataset/file_csv.cc
Comment (westonpace) [inline cpp/src/arrow/dataset/file_csv.cc:204]:
```suggestion
  // If the file is uncompressed we open the reader with a RandomAccessFile which will
  // be capable of reading the file in parallel.  If the file is compressed we must use an
  // input stream and will be read sequentially.
  if (actual_compression == Compression::type::UNCOMPRESSED) {
```

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:977]:
```suggestion
```

**Review by westonpace — APPROVED** [review]
I have a few minor comment / wording suggestions but otherwise I think this is good to go.

**Comment by marsupialtail** [comment]
@lidavidm what do you think I should do with StreamWrapFunc?

**Comment by lidavidm** [comment]
@marsupialtail from a quick scan - yeah, input transcoding would disable parallelism just like compression. I'd probably not worry about it: it already means we have to call back into Python to read data.

**Comment by marsupialtail** [comment]
OK this won't be supported in fast CSV reader then

**Comment by westonpace** [comment]
I agree that we should leave it for now.  In the future we should still be able to do parallel I/O, we would just need to have a serial compute path on top of it.  I suppose we could do something like `RandomAccessFile -> Generator<Buffer> -> Readahead -> Iterator<Buffer> -> InputStream` :laughing: but maybe we could find some solution with fewer layers too.

**Comment by marsupialtail** [comment]
I think the test errors are spurious. Should be ready to be merged.

File: cpp/src/arrow/io/interfaces.h
Comment (pitrou) [inline cpp/src/arrow/io/interfaces.h:27]: This is a too heavy inclusion, should instead just use the corresponding `std::function<...>` definition.

File: cpp/src/arrow/csv/reader.cc
Comment (pitrou) [inline cpp/src/arrow/csv/reader.cc:975]: I'm not found of all the duplication here. How about having `StreamingReaderImpl` take a buffer-async-generator instead?

File: cpp/src/arrow/io/interfaces.h
Comment (pitrou) [inline cpp/src/arrow/io/interfaces.h:348]: If this is beneficial for reading from S3, should it perhaps be called in S3 and GCS filesystem implementations of `OpenInputStream` instead of just the CSV datasets reader?
(does that sound doable/reasonable?)

File: cpp/src/arrow/io/interfaces.h
Comment (pitrou) [reply cpp/src/arrow/io/interfaces.h:27]: Not in this `.h`, only in the `.cc`

Comment (marsupialtail) [reply cpp/src/arrow/io/interfaces.h:27]: Ah I see

File: cpp/src/arrow/csv/reader.cc
Comment (marsupialtail) [reply cpp/src/arrow/csv/reader.cc:975]: This code duplication is admittedly not ideal. However it's not just about reading faster through a new RandomAccessFile generator, but also parsing in parallel, and (in the future) decoding in parallel, which will introduce more changes to the StreamingReaderImpl. 

After I finish the decoder in parallel in a new PR perhaps then it makes sense to reduce the deduplication. Or perhaps I can just add it to this PR, though I think this PR has enough value to be merged by itself.

Comment (marsupialtail) [reply cpp/src/arrow/csv/reader.cc]: This has been addressed.

**Comment by marsupialtail** [comment]
Should be good now

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:888]: Are these sorts of changes (where you are adding `ReaderMixin<InputType>::` needed to resolve some kind of ambiguity?  I don't know that we have any specific stance in the style guide so I'm not sure it's a problem but I'm curious why you made the change.

**Review by westonpace — APPROVED** [review]
This removes the duplication enough to satisfy my concerns.

File: cpp/src/arrow/csv/reader.cc
Comment (marsupialtail) [reply cpp/src/arrow/csv/reader.cc:888]: Needs the qualifiers to compile

**Comment by westonpace** [comment]
Can you fix the appveyor failure?  I suspect you just need to update the test case to the new approach.

**Comment by marsupialtail** [comment]
Need to keep up to date with the times (C++17)

**Comment by marsupialtail** [comment]
rebased on master

**Comment by marsupialtail** [comment]
Messed up the rebase, made a new branch: https://github.com/apache/arrow/pull/14269

**Pull request closed without merging**
